include/arch/arm/aarch32/syscall.h

```diff
@@ -163,19 +165,37 @@ static inline uintptr_t arch_syscall_invoke0(uintptr_t call_id)
 	return ret;
 }
 
+static inline bool aarch32_is_unprivileged_state(void)
+{
+	uint32_t value;
+
+#if defined(CONFIG_CPU_CORTEX_R)
+	__asm__ volatile("mrs %0, CPSR\n\t" : "=r"(value));
+
+	/*
+	 * For Cortex-R, the mode bits (lower 5 bits) will be 0x10 for user mode.
+	 */
+	return ((value & 0x1f) == 0x10) ? true : false;
+#else
+	/* return mode information */
+	__asm__ volatile("mrs %0, CONTROL\n\t" : "=r"(value));
+	return (value & CONTROL_nPRIV_Msk) ? true : false;
+#endif
+}
+
 static inline bool arch_is_user_context(void)
 {
+#if defined(CONFIG_CPU_CORTEX_M)
 	uint32_t value;
 
 	/* check for handler mode */
 	__asm__ volatile("mrs %0, IPSR\n\t" : "=r"(value));
 	if (value) {
 		return false;
 	}
+#endif
 
-	/* if not handler mode, return mode information */
-	__asm__ volatile("mrs %0, CONTROL\n\t" : "=r"(value));
-	return (value & CONTROL_nPRIV_Msk) ? true : false;
+	return aarch32_is_unprivileged_state();
```

Comment:
This is fine (see, though my comment about the function name)

Reply:
fixed